runtime/src/commonMain/kotlin/pbandk/internal/TimeUtil.kt

```diff
@@ -160,36 +138,36 @@ internal fun formatTime(seconds: Long, nanos: Int): String {
     }
     val time = DateTime.fromSeconds(seconds)
 
-    return memScoped {
-        // The output string will be yyyy-mm-ddThh:mm:ss.sssssssssZ. The fractional seconds component (with
-        // leading period) is optional. If present, it can be up to 9 digits long but can be shorter.
-        val outputBufferSize = 40
-        val output = allocArray<ByteVar>(outputBufferSize)
-        var bytesWritten = snprintf(
-            output, outputBufferSize.convert(), "%04d-%02d-%02dT%02d:%02d:%02d",
-            time.year, time.month, time.day,
-            time.hour, time.minute, time.second
-        )
-        if (bytesWritten < 0) {
-            throw PosixException(posix_errno())
-        } else if (bytesWritten > outputBufferSize) {
-            throw IllegalStateException("Needed to write $bytesWritten bytes but only had space for $outputBufferSize")
-        }
-
+    return buildString(40) {
+        append(time.year.toString().padStart(4, '0'))
+        append('-')
+        append(time.month.toString().padStart(2, '0'))
+        append('-')
+        append(time.day.toString().padStart(2, '0'))
+        append('T')
+        append(time.hour.toString().padStart(2, '0'))
+        append(':')
+        append(time.minute.toString().padStart(2, '0'))
+        append(':')
+        append(time.second.toString().padStart(2, '0'))
         if (nanos != 0) {
-            output.setChecked(bytesWritten++, outputBufferSize, '.')
-            check(bytesWritten + 9 < outputBufferSize) {
-                "Not enough space in buffer to write nanos: index=$bytesWritten, size=$outputBufferSize"
-            }
-            bytesWritten += formatNanos((output + bytesWritten.toLong())!!, nanos)
+            formatNanos(this, nanos)
```

Comment:
This is a total nitpick, but can you keep the addition of the `'.'` character in here rather than moving it to `formatNanos()`? I'm trying to keep this file as close as possible in structure to the `time.cc` file from the protobuf C++ library that this code was copied from. That makes it easier to compare the two files.

Reply:
Oh my goodness why didn't I think of that... I was looking for a place to put the "." so I only had to use it once.